app/repos/activity.js

```diff
@@ -52,7 +56,92 @@ const getActivityById = async (id) => {
   }
 }
 
+/**
+ * @typedef ActivityCreatePayload
+ * @property {string} label
+ * @property {string} activityType
+ * @property {string} activitySource
+ */
+/**
+ * @param {ActivityCreatePayload} activityData
+ * @param user
+ * @param [transaction]
+ * @returns {Promise<*|undefined>}
+ */
+const createActivity = async (activityData, user, transaction) => {
+  if (!transaction) {
+    return await sequelize.transaction(async (t) => createActivity(activityData, user, t))
+  }
+
+  const activityTypes = await getActivityList(activityData.activityType, activityData.activitySource)
+
+  const foundActivities = activityTypes.filter(x => x.label.toLowerCase() === activityData.label.toLowerCase())
+  const foundActivity = foundActivities?.length > 0 ? foundActivities[0] : null
+
+  if (foundActivity !== null) {
+    throw new DuplicateResourceError(`Activity with name ${activityData.label} already exists for type ${activityData.activityType} and source ${activityData.activitySource}`)
+  }
+
+  const activitySource = await sequelize.models.activity_source.findOne({
+    where: { name: activityData.activitySource }
+  })
+
+  const activityType = await sequelize.models.activity_type.findOne({
+    where: { name: activityData.activityType }
+  })
+
+  const activity = await sequelize.models.activity.create({
+    label: activityData.label,
+    activity_source_id: activitySource.id,
+    activity_type_id: activityType.id,
+    activity_event_id: activitySource.id,
+    display_order: 10
+  }, { transaction })
+
```

Comment:
not sure what the plan is on soft-deletion, but if we do add it, we will need some code here to restore soft deleted activities.

Reply:
No soft-deletion needed. There is no referential integrity since the results are records in the events servie, not in the DB. The DB records only drive what can be shown on the 'add activity' screens